src/date.c

```diff
@@ -1,18 +1,983 @@
-  struct v7_val *obj = cfa->called_as_constructor ? cfa->this_obj :
-                       v7_push_new_object(cfa->v7);
-  v7_set_class(obj, V7_CLASS_DATE);
-  return V7_OK;
+#ifdef __APPLE__
+int64_t strtoll(const char *, char **, int);
+#endif
+
+extern long timezone;
+
+typedef double etime_t; /* double is suitable type for ECMA time */
+/* inernally we have to use 64-bit integer for some operations */
+typedef int64_t etimeint_t;
+#define INVALID_TIME NAN
+
+/*++++++  ECMA date & time helpers ++++++*/
+
+#define msPerDay 86400000
+#define HoursPerDay 24
+#define MinutesPerHour 60
+#define SecondsPerMinute 60
+#define SecondsPerHour 3600
+#define msPerSecond 1000
+#define msPerMinute 60000
+#define msPerHour 3600000
+#define MonthsInYear 12
+
+static etimeint_t ecma_Day(etime_t t) {
+  return floor(t/ msPerDay);
+}
+
+static etimeint_t gmtoffms; /* timezone offset, ms, no DST */
+
+/* Leap year formula copied from ECMA 5.1 standart as is */
+static int ecma_DaysInYear(int y) {
+  if(y % 4 != 0 ) {
+    return 365;
+  } else if( y % 4 == 0 && y % 100 != 0 ) {
+    return 366;
+  } else if(y % 100 == 0 && y % 400 != 0) {
+    return 365;
+  } else if(y % 400 == 0) {
+    return 366;
+  } else {
+    return 365;
+  }
+}
+
+static etimeint_t ecma_DayFromYear(etimeint_t y) {
+  return 365 * (y-1970) + floor((y-1969)/4) - floor((y-1901)/100)
+                                              + floor((y-1601)/400);
+}
+
+static etimeint_t ecma_TimeFromYear(etimeint_t y) {
+  return msPerDay * ecma_DayFromYear(y);
+}
+
+static int ecma_YearFromTime_s(etime_t t)
+{
+  int first = floor((t / msPerDay) / 366) + 1970,
+              last = floor((t / msPerDay) / 365) + 1970,
+              middle = 0;
+  
+  if (last < first) {
+    int temp = first;
+    first = last;
+    last = temp;
+  }
+  
+  while (last > first) {
+    middle = (last + first) / 2;
+    if (ecma_TimeFromYear(middle) > t) {
+      last = middle - 1;
+    } else {
+      if (ecma_TimeFromYear(middle) <= t) {
+        if (ecma_TimeFromYear(middle + 1) > t) {
+          first = middle;
+          break;
+        }
+        first = middle + 1;
+      }
+    }
+  }
+  
+  return first;
+}
+
+static int ecma_InLeapYear(etime_t t, int year) {
+  (void)t;
+  return ecma_DaysInYear(year) == 366;
+}
+
+static int ecma_DayWithinYear(etime_t t, int year) {
+  return (int)(ecma_Day(t) - ecma_DayFromYear(year));
+}
+
+static int* ecma_getfirstdays(int isleap) {
+  static int sdays[2][MonthsInYear+1] =
+    {{0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334, 365},
+     {0, 31, 60, 91, 121, 152, 182, 213, 244, 274, 305, 335, 366}};
+  
+  return sdays[isleap];
+}
+
+static int ecma_MonthFromTime(etime_t t, int year) {
+  int *days;
+  etimeint_t dwy = ecma_DayWithinYear(t, year);
+  
+  unsigned int i; int ret = -1;
+  
+  days = ecma_getfirstdays(ecma_InLeapYear(t, year));
+  
+  for(i = 0; i < MonthsInYear; i++) {
+    if(dwy >= days[i] && dwy < days[i+1]) {
+      ret = i;
```

Comment:
Both approaches have their advantages.
We don't have specific fine grained rules about it; the general idea is to not create complicated control flow just to abide to some kind of abstract structured programming practices (unless they buy you something tangible).

In this case, given the nesting level, I find the extra brain cycles needed to check whether this break actually breaks the loop or if there is another loop to not be worth. And the less variables the better.

In other places we even use gotos to jump to common cleanup code.